### PR TITLE
Implement package assembly in scancode.io

### PR DESCRIPTION
**Pull request by JonoYang**
This PR updates the application package scanning step in scancode.io to reflect the updates made to package scanning in scancode-toolkit. Package scanning is now a two step process, where we first detect Package data from CodebaseResources, then we iterate through all the Resources with Package data and process them using the new Package handlers from scancode-toolkit.

### CODE REVIEW

**Comment by tdruez** [comment]
@JonoYang thanks for splitting those PR, much easier to review and test.

1. I'm getting the following error running the scan_codebase on https://files.pythonhosted.org/packages/9b/41/e1e7d6ecc3bc76681dfdc6b373566822bc2aab96fa3eceaaed70accc28b6/Django-4.1-py3-none-any.whl

```
DiscoveredPackage matching query does not exist.

Traceback:
  File "scancode.io/scanpipe/pipelines/__init__.py", line 115, in execute
    step(self)
  File "scancode.io/scanpipe/pipelines/scan_codebase.py", line 93, in scan_for_application_packages
    scancode.scan_for_application_packages(self.project)
  File "scancode.io/scanpipe/pipes/scancode.py", line 332, in scan_for_application_packages
    assemble_packages(project=project)
  File "scancode.io/scanpipe/pipes/scancode.py", line 375, in assemble_packages
    for item in items:
  File "scancode.io/lib/python3.9/site-packages/packagedcode/models.py", line 941, in assemble
    cls.assign_package_to_resources(
  File "scancode.io/lib/python3.9/site-packages/packagedcode/models.py", line 1005, in assign_package_to_resources
    package_adder(package_uid, resource, codebase)
  File "scancode.io/scanpipe/pipes/scancode.py", line 344, in add_to_package
    package = project.discoveredpackages.get(package_uid=package_uid)
  File "scancode.io/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "scancode.io/lib/python3.9/site-packages/django/db/models/query.py", line 496, in get
    raise self.model.DoesNotExist(
```

2. We should add a simple test for assemble_packages, to expose how a couple resources with package_data are assembled into a package.

3. We need to optimize this data migration https://github.com/nexB/scancode.io/pull/485/files#diff-53edf14fc544b3829f4f4afbe3b0e97534e97829e9bc3d1a9be7e4e2680868b0R6
I just takes too much time on large database with hundred of thousands existing resources. Loading the resources then calling save() on each one is too slow, I would suggested to use update() instead. The goal is to convert all the Python logic into SQL logic. We may need to write raw SQL here if the Django QuerySet API is not enough.

**Comment by tdruez** [comment]
@JonoYang Did you have the chance to investigate the issue reported above?

**Comment by JonoYang** [comment]
@tdruez The issue is caused by `DatafileHandler.assemble_from_many()`, where Resources are yielded before Packages. I've set up the CodebaseResource creation logic in `load_inventory` and `scan_codebase` to create `DiscoveredPackages` before the `CodebaseResources` of a Project, such that I can associate the `CodebaseResources` to their `DiscoveredPackages`. I've modified the `.assemble()` methods of other DatafileHandlers to yield Packages before dependencies and resources, but I'm also looking into sorting these objects on the scancode.io side in the `assemble_packages` function in my PR.

This is my PR on scancode-toolkit that modifies `DatafileHandler.assemble_from_many()` to yield Packages before Resources and Dependencies
https://github.com/nexB/scancode-toolkit/pull/3042

**Comment by JonoYang** [comment]
Sorting the items yielded by `.assemble()` doesn't work. We still run into the issue where we need a DiscoveredPackage to exist for a Project before we can relate a CodebaseResource to it when the `.assemble()` method code is determining the resources of a package.

**Comment by tdruez** [comment]
@JonoYang could you extract the changes related to https://github.com/nexB/scancode.io/issues/467 into it's own PR so we can get this one merge while we are sorting the assembly issues?

**Comment by JonoYang** [comment]
@tdruez 

I've updated `CodebaseResource.for_packages` to return a list of `package_uid`s of the DiscoveredPackages associated to the CodebaseResource. `package_uid` values are different every time they are generated, so in tests where we compare `package_uid` values, we normalize them using `purl_with_fake_uuid`. I was wondering what was the best way to normalize a `package_uid` from the CSV output test (https://github.com/nexB/scancode.io/blob/implement-package-assembly/scanpipe/tests/test_pipes.py#L132)

**Comment by JonoYang** [comment]
> I've updated CodebaseResource.for_packages to return a list of package_uids of the DiscoveredPackages associated to the CodebaseResource. package_uid values are different every time they are generated, so in tests where we compare package_uid values, we normalize them using purl_with_fake_uuid. I was wondering what was the best way to normalize a package_uid from the CSV output test (https://github.com/nexB/scancode.io/blob/implement-package-assembly/scanpipe/tests/test_pipes.py#L132)

Looking at it again, I now realize there is no need to normalize the` package_uid`s in the `for_packages` field in the CSV tests because the Package data is set in code and in the fixtures and we will always get the same `package_uid` returned. In scancode-toolkit, we need to normalize the `package_uid`s because scancode generates a unique `package_uid` on each test run.

File: scanpipe/models.py [hunk 1]
```diff
@@ -1248,6 +1257,9 @@ def has_licenses(self):
     def has_no_licenses(self):
         return self.filter(licenses=[])
 
+    def has_package_data(self):
+        return self.filter(package_data__isnull=False)
```
Comment (tdruez) [inline scanpipe/models.py:1261]: `isnull` is not the proper lookup to use for this field that has an empty list a default value.

```
self.filter(~Q(package_data=[]))
```

**Review by tdruez — CHANGES_REQUESTED** [review]

**Comment by tdruez** [comment]
@JonoYang I'm not sure your latest changes were intended to fix https://github.com/nexB/scancode.io/pull/485#issuecomment-1206381058 but I'm still getting the same issue with the latest code.

**Comment by JonoYang** [comment]
> @JonoYang I'm not sure your latest changes were intended to fix [#485 (comment)](https://github.com/nexB/scancode.io/pull/485#issuecomment-1206381058) but I'm still getting the same issue with the latest code.

The PR with this fix has been merged into scancode-toolkit and should be available in the next version and I'll bump the scancode version in scio when it comes out. (https://github.com/nexB/scancode-toolkit/pull/3042)

**Comment by tdruez** [comment]
@JonoYang I've merged the latest main branch including the ScanCode-toolkit v31.0.1 version.

1. The issue "DiscoveredPackage matching query does not exist." from https://github.com/nexB/scancode.io/pull/485#issuecomment-1206381058 appears to be fixed.

I'm seeing an extra resource associated to the second package using the main branch that I do not see in this current branch:

* `main`
Screenshot 2022-08-18 at 15 15 14

* `implement-package-assembly`
Screenshot 2022-08-18 at 15 16 03

Also, shouldn't we only detect 1 package here?

----- 

2. I'm still getting this issue in the `assemble_packages()` with this branch running the `docker` pipeline on `docker://python:3.8-slim-buster`:

```
CodebaseResource matching query does not exist.

Traceback:
  File "/app/scanpipe/pipelines/__init__.py", line 115, in execute
    step(self)
  File "/app/scanpipe/pipelines/root_filesystems.py", line 118, in scan_for_application_packages
    scancode.scan_for_application_packages(self.project)
  File "/app/scanpipe/pipes/scancode.py", line 332, in scan_for_application_packages
    assemble_packages(project=project)
  File "/app/scanpipe/pipes/scancode.py", line 375, in assemble_packages
    for item in items:
  File "/usr/local/lib/python3.9/site-packages/packagedcode/models.py", line 949, in assemble
    cls.assign_package_to_resources(
  File "/usr/local/lib/python3.9/site-packages/packagedcode/pypi.py", line 319, in assign_package_to_resources
    site_packages = resource.parent(codebase).parent(codebase).parent(codebase)
  File "/app/scanpipe/models.py", line 1632, in parent
    return parent_path and self.project.codebaseresources.get(path=parent_path)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 650, in get
    raise self.model.DoesNotExist(
```

**Comment by JonoYang** [comment]
@tdruez 

> CodebaseResource matching query does not exist.
> 
> Traceback:
>   File "/app/scanpipe/pipelines/__init__.py", line 115, in execute
>     step(self)
>   File "/app/scanpipe/pipelines/root_filesystems.py", line 118, in scan_for_application_packages
>     scancode.scan_for_application_packages(self.project)
>   File "/app/scanpipe/pipes/scancode.py", line 332, in scan_for_application_packages
>     assemble_packages(project=project)
>   File "/app/scanpipe/pipes/scancode.py", line 375, in assemble_packages
>     for item in items:
>   File "/usr/local/lib/python3.9/site-packages/packagedcode/models.py", line 949, in assemble
>     cls.assign_package_to_resources(
>   File "/usr/local/lib/python3.9/site-packages/packagedcode/pypi.py", line 319, in assign_package_to_resources
>     site_packages = resource.parent(codebase).parent(codebase).parent(codebase)
>   File "/app/scanpipe/models.py", line 1632, in parent
>     return parent_path and self.project.codebaseresources.get(path=parent_path)
>   File "/usr/local/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
>     return getattr(self.get_queryset(), name)(*args, **kwargs)
>   File "/usr/local/lib/python3.9/site-packages/django/db/models/query.py", line 650, in get
>     raise self.model.DoesNotExist(

This error is happening because we do not create the directory Resources when we create Resources from a docker image. The fix for this is to set `with_dir=True` as an argument into `layer.get_resources()` (https://github.com/nexB/scancode.io/blob/main/scanpipe/pipes/docker.py#L176)

After fixing that issue, the pipeline failed from a syntax error in `PythonInstalledWheelMetadataFile.assign_package_to_resources()`. I've created a PR for it (https://github.com/nexB/scancode-toolkit/pull/3059)

After fixing the syntax error in `PythonInstalledWheelMetadataFile.assign_package_to_resources()`, I'm able to run the docker pipeline without errors.
image

**Comment by tdruez** [comment]
@JonoYang with the latest toolkit 31.0.2 the DoesNotExist is now properly fixed.
I've encountered a new ProjectError reported https://github.com/nexB/scancode-toolkit/issues/3067 but unrelated to this PR.
Merging!

**Pull request merged**